app/models/server.rb

```diff
@@ -27,4 +32,102 @@ class Server < ApplicationRecord
     with: /([0-9A-Fa-f]{2}[:-]){5}([0-9A-Fa-f]{2})/,
     message: 'is not a valid MAC address'
   }
+
+  after_commit :save_users
+  after_initialize :assign_defaults
+
+  attr_writer :sudo_users
+  attr_writer :users
+
+  def user_ids=(new_user_ids)
+    @users = User.find(new_user_ids)
+  end
+
+  def sudo_user_ids=(new_sudo_users)
+    @sudo_users = User.find(new_sudo_users)
+  end
+
+  # in this model, the groups: users, sudo_users, and responsible_users shall be seperate groups
+  # They will be merged for writing puppet scripts and seperated when reading them
+
+  def sudo_users
+    read_users if @sudo_users.nil?
+
+    @sudo_users
+  end
+
+  def users
+    read_users if @users.nil?
+
+    @users
+  end
+
+  def all_users
+    (users + sudo_users + responsible_users).uniq
+  end
+
+  # This makes a few things easier, for example when adding arrays together
+  def responsible_users
+    responsible.nil? ? [] : [responsible]
```

Comment:
What is repsonsible? If it is an array there is an array in an array. Is that intended?